File: test/language/templates/java/templates/InstantiateTypeAsParameterTest.java
package templates;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;

import org.junit.jupiter.api.Test;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

import templates.instantiate_type_as_parameter.InstantiateTypeAsParameter;
import templates.instantiate_type_as_parameter.P32;
import templates.instantiate_type_as_parameter.Parameterized_P32;

public class InstantiateTypeAsParameterTest
{
    @Test
    public void readWrite() throws IOException
    {
        final P32 param = new P32(2);
        final long[] array = new long[] {13, 42};
        final Parameterized_P32 parameterized = new Parameterized_P32(param, array);

        final InstantiateTypeAsParameter instantiateTypeAsParameter =
                new InstantiateTypeAsParameter(param, parameterized);

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        instantiateTypeAsParameter.write(writer);

        final BitStreamReader reader =
                new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        InstantiateTypeAsParameter readInstantiateTypeAsParameter = new InstantiateTypeAsParameter(reader);
        assertTrue(instantiateTypeAsParameter.equals(readInstantiateTypeAsParameter));
    }
}

File: test/extensions/language/templates/python/InstantiateTypeAsParameterTest.py
import zserio

import Templates


class InstantiateTypeAsParameterTest(Templates.TestCase):
    def testReadWrite(self):
        param = self.api.P32(2)
        instantiateTypeAsParameter = self.api.InstantiateTypeAsParameter(
            param, self.api.Parameterized_P32(param, [13, 42])
        )

        writer = zserio.BitStreamWriter()
        instantiateTypeAsParameter.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readInstantiateTypeAsParameter = self.api.InstantiateTypeAsParameter()
        readInstantiateTypeAsParameter.read(reader)
        self.assertEqual(instantiateTypeAsParameter, readInstantiateTypeAsParameter)
